src/util/strconv.h

```diff
@@ -106,11 +110,97 @@ static char *rm_strndup_unescape(const char *s, size_t len) {
   return ret;
 }
 
+// transform utf8 string to lowercase using nunicode library
+// encoded: the utf8 string to transform
+// in_len: the length of the utf8 string
+// dst: the destination buffer to store the transformed string
+// returns the bytes written to encoded, or 0 if the length of the transformed
+// string is greater than in_len and no transformation was done
+static size_t nunicode_tolower(const char *encoded, size_t in_len, char *dst) {
+  char encoded_stack_buffer[SSO_MAX_LENGTH];
+  const char *encoded_copy = encoded_stack_buffer;
+  uint32_t u_stack_buffer[SSO_MAX_LENGTH * 2];
+  uint32_t *u_buffer = u_stack_buffer;
+
+  if (in_len == 0) {
+    return 0;
+  }
+
+  // create a zero-terminated copy of the encoded string
+  if (in_len >= SSO_MAX_LENGTH - 1) {
+    encoded_copy = rm_strndup(encoded, in_len);
+  } else {
+    // do not use heap allocation for short strings
+    memcpy(encoded_stack_buffer, encoded, in_len);
+    encoded_stack_buffer[in_len] = '\0';
+  }
+
+  const char *encoded_char = encoded_copy;
+  ssize_t u_len = nu_strtransformnlen(encoded_copy, in_len, nu_utf8_read,
+                                              nu_tolower, nu_casemap_read);
+
+  if (u_len >= (SSO_MAX_LENGTH * 2) - 1) {
+    u_buffer = (uint32_t *)rm_malloc(sizeof(*u_buffer) * (u_len + 1));
+  }
+
+  // Decode utf8 string into Unicode codepoints and convert to lowercase
+  uint32_t codepoint;
+  unsigned i = 0;
+  while (1) {
+    // Read unicode codepoint from utf8 string
+    encoded_char = nu_utf8_read(encoded_char, &codepoint);
+    // Transform unicode codepoint to lowercase
+    const char *map = nu_tolower(codepoint);
+
+    // Read the transformed codepoint and store it in the unicode buffer
+    if (map != 0) {
+      uint32_t mu;
+      while (1) {
+        map = nu_casemap_read(map, &mu);
+        if (mu == 0) {
+          break;
+        }
+        u_buffer[i] = mu;
+        ++i;
+      }
+    }
+    else {
+      // If no transformation is needed, just copy the unicode codepoint
+      u_buffer[i] = codepoint;
+      ++i;
+    }
+
+    // Break if the end of the string is reached
+    if (codepoint == 0) {
+      break;
+    }
+  }
+
+  // Encode Unicode codepoints back to utf8 string
+  ssize_t reencoded_len = nu_bytelen(u_buffer, nu_utf8_write);
```

Comment:
In sample (https://bitbucket.org/alekseyt/nunicode/src/master/samples/folding.c)
it is not done (using `nu_utf8_write`)

Reply:
I think we need to write to the intermediate unicode buffer `u_buffer` because we need to check that we don't exceed the length of the original `utf8` string